### ISSUE TITLE
Convolutional Output Layers?

### ISSUE THREAD

**Issue by bloc97** [deeplearning4j/deeplearning4j#3614]
#### Issue Description

I cannot find in the documentation how to add a Convolutional Output Layer.
The model's output is a 3D tensor, and the loss function should be MSE.

Do I have to implement it myself using `IOutputLayer` or does this already exist in the library?

Update: I have tried the `LossLayer` layer, but it seems that this layer ignores the last dimension of the 3D tensor, which gives an error.
 `Exception in thread "main" java.lang.IllegalStateException: Mis matched lengths: [192] != [24576]`

Here is the code.

        double learningRate = 0.1d;
        
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(182)
                .iterations(10)
                .learningRate(learningRate)
                .gradientNormalization(GradientNormalization.ClipElementWiseAbsoluteValue)
                .gradientNormalizationThreshold(0.001d * learningRate)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .regularization(true)
                .l2(0.0001d) //Weight decay
                .updater(Updater.NESTEROVS)
                .momentum(0.9d)
                .weightInit(WeightInit.XAVIER)
                .list()
                .layer(0, new ConvolutionLayer.Builder()
                        .nIn(3)
                        .kernelSize(3, 3)
                        .stride(1, 1)
                        .padding(1, 1)
                        .nOut(3)
                        .activation(new ActivationELU(0.3d))
                        .build())
                .layer(1, new LossLayer.Builder(LossFunctions.LossFunction.MSE).build())
                .backprop(true).pretrain(false).build();
        
        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.init();
        
        INDArray inputs = Nd4j.zeros(2, 3, 64, 64);
        INDArray labels = Nd4j.zeros(2, 3, 64, 64);
        
        System.out.println(model.output(inputs).length());
        
        for( int i=0; i<10; i++ ) {
            model.fit(inputs, labels);
            System.out.println("*** Completed epoch {" + i + "} ***");
        }

**Comment by hiiroo** [comment]
I don't know if you can set output layers by names in MultiLayerConfiguration but you can set multiple output points in a ComputationGraph using `.setOutput("NameOfTheLayer")`. Here is an example code I have tried before. It outputs "convolution1" layer.

```java


ComputationGraphConfiguration conf = new NeuralNetConfiguration.Builder()
    .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).iterations(1)
    .learningRate(0.1)
    .seed(12345)
    .regularization(true)
    .l2(0.001)
    .weightInit(WeightInit.XAVIER)
    .graphBuilder()
    .addInputs("input")
    .setInputTypes(InputType.convolutionalFlat(28, 28, 1))
    .addLayer("convolution1", new ConvolutionLayer.Builder(4, 4)
        .nIn(nChannels)
        .stride(2, 2)
        .padding(0, 0)
        .nOut(20)
        .activation("relu")
        .build(), "input")
    .setOutputs("convolution1")
    .backpropType(BackpropType.Standard)
    .pretrain(false).backprop(true)
    .build();

ComputationGraph net = new ComputationGraph(conf);
net.init();

```

**Comment by agibsonccc** [comment]
You need an ouput layer of some kind on the end. You'll notice the lack of a loss function definition on the conv layer. You need that for backprop. 
We have an rnn output layer that assumes 3d.
What exactly is your use case?

**Comment by AlexDBlack** [comment]
Right, it's not currently supported. Probably the best approach here would be to add another output layer that does reshaping internally, before passing the (now 2d) values to the loss function. That's how RnnOutputLayer works, and we should be able to do the same thing for CNN outputs.

**Comment by eraly** [comment]
@agibsonccc @AlexDBlack Would we want the user to specify height and width in the builder? I guess .nin would be the depth? Would this essentially be the same as a dense output layer with nIn(hxwxd)?

**Comment by AlexDBlack** [comment]
If we add a CnnOutputLayer, then the user should not have to provide any dimensions in the config. In fact, there are some use cases where the output dimensions (height/width) will vary for different inputs.

How I imagine this CnnOutputLayer would work:
- The input activations to the CnnOutputLayer should be standard 4d image layout: ```[minibatch, depth, height, width]```
- The labels would be 4d, same shape as input
- For reshaping purposes, we'd reshape to 2d to ```[minibatch * height * width, depth]```
- Consequently, for things like softmax+mcxent and the like: we'd essentially be treating the 'depth' dimension as the number of classes; the softmax would also be applied along dimension 1 (same as OutputLayer/RnnOutputLayer currently does). And like with RnnOutputLayer + time/examples, we treat each of the ```minibatch x height x width``` inputs separately
- Only complication here: per-example masking and per-output masking.
    - Per-output masking should be easy enough (same reshaping procedure as the labels/activations)
    - Per-*example* masking would need extra thought (we need a larger column vector with values that match the post-reshaped activations/labels)

**Comment by AlexDBlack** [comment]
Added here: will be merged soon. https://github.com/deeplearning4j/deeplearning4j/pull/4389
